AK/SIMDExtras.h

```diff
@@ -0,0 +1,158 @@
+
+// SIMD Vector Expansion
+
+ALWAYS_INLINE static constexpr f32x4 expand4(float f)
+{
+    return f32x4 { f, f, f, f };
+}
+
+ALWAYS_INLINE static constexpr i32x4 expand4(i32 i)
+{
+    return i32x4 { i, i, i, i };
+}
+
+ALWAYS_INLINE static constexpr u32x4 expand4(u32 u)
+{
+    return u32x4 { u, u, u, u };
+}
+
+// Casting
+
+template<typename TSrc>
+ALWAYS_INLINE static u32x4 to_u32x4(TSrc v)
+{
+    return __builtin_convertvector(v, u32x4);
+}
+
+template<typename TSrc>
+ALWAYS_INLINE static i32x4 to_i32x4(TSrc v)
+{
+    return __builtin_convertvector(v, i32x4);
+}
+
+template<typename TSrc>
+ALWAYS_INLINE static f32x4 to_f32x4(TSrc v)
+{
+    return __builtin_convertvector(v, f32x4);
+}
+
+// Rounding
+
+ALWAYS_INLINE static f32x4 truncate(f32x4 v)
+{
+    return to_f32x4(to_i32x4(v));
+}
+
+ALWAYS_INLINE static f32x4 floor(f32x4 v)
+{
+    v = (v < 0.f) ? v - 1 : v;
+    return truncate(v);
+}
+
+ALWAYS_INLINE static f32x4 ceil(f32x4 v)
+{
+    v = (v >= 0.f) ? v + 1 : v;
+    return truncate(v);
+}
+
+ALWAYS_INLINE static f32x4 frac(f32x4 v)
+{
+    return v - floor(v);
+}
+
+ALWAYS_INLINE static f32x4 clamp(f32x4 v, f32x4 min, f32x4 max)
+{
+    return v < min ? min : (v > max ? max : v);
+}
+
+// Masking
+
+ALWAYS_INLINE static i32 maskbits(i32x4 mask)
+{
+    return __builtin_ia32_movmskps((f32x4)mask);
+}
+
+ALWAYS_INLINE static bool all(i32x4 mask)
+{
+    return maskbits(mask) == 15;
+}
+
+ALWAYS_INLINE static bool any(i32x4 mask)
+{
+    return maskbits(mask) != 0;
+}
+
+ALWAYS_INLINE static bool none(i32x4 mask)
+{
+    return maskbits(mask) == 0;
+}
+
+ALWAYS_INLINE static int maskcount(i32x4 mask)
+{
+    constexpr static int count_lut[16] { 0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4 };
+    return count_lut[maskbits(mask)];
+}
+
+// Load / Store
+
+ALWAYS_INLINE static f32x4 load4(float const* a, float const* b, float const* c, float const* d)
+{
+    return f32x4 { *a, *b, *c, *d };
+}
+
+ALWAYS_INLINE static void store4(f32x4 v, float* a, float* b, float* c, float* d)
+{
+    *a = v[0];
+    *b = v[1];
+    *c = v[2];
+    *d = v[3];
+}
+
+ALWAYS_INLINE static f32x4 load4_masked(float const* a, float const* b, float const* c, float const* d, i32x4 mask)
+{
+    int bits = maskbits(mask);
+    return f32x4 {
+        bits & 1 ? *a : 0.f,
+        bits & 2 ? *b : 0.f,
+        bits & 4 ? *c : 0.f,
+        bits & 8 ? *d : 0.f,
+    };
+}
+
+ALWAYS_INLINE static void store4_masked(f32x4 v, float* a, float* b, float* c, float* d, i32x4 mask)
+{
+    int bits = maskbits(mask);
+    if (bits & 1)
+        *a = v[0];
+    if (bits & 2)
+        *b = v[1];
+    if (bits & 4)
+        *c = v[2];
+    if (bits & 8)
+        *d = v[3];
+}
+
+ALWAYS_INLINE static void store4_masked(u32x4 v, u32* a, u32* b, u32* c, u32* d, i32x4 mask)
+{
+    int bits = maskbits(mask);
+    if (bits & 1)
+        *a = v[0];
+    if (bits & 2)
+        *b = v[1];
+    if (bits & 4)
+        *c = v[2];
+    if (bits & 8)
+        *d = v[3];
+}
```

Comment:
this could be a template:
```suggestion
template<typename VectorType, typename UnderlyingType = decltype(declval<VectorType>[0])>
ALWAYS_INLINE static void store4_masked(VectorType v, UnderlyingType * a, UnderlyingType* b, UnderlyingType* c, UnderlyingType* d, i32x4 mask)
{
    int bits = maskbits(mask);
    if (bits & 1)
        *a = v[0];
    if (bits & 2)
        *b = v[1];
    if (bits & 4)
        *c = v[2];
    if (bits & 8)
        *d = v[3];
}
```